pkg/csi_driver/controller.go

```diff
@@ -101,6 +104,30 @@ func (s *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVolu
 			return nil, status.Error(codes.AlreadyExists, err.Error())
 		}
 	} else {
+
+		// Get the reserved-IPV4-cidr from the request. If value is not provided, reservedIPV4Range is defaulted to ""
+		reservedIPV4CIDR, ok := req.GetParameters()[paramReservedIPV4CIDR]
+		if ok {
+
+			// Prevents lock acquisition in case the CIDR is invalid
+			validCIDR := s.config.ipAllocator.ValidateCIDR(reservedIPV4CIDR)
+			if !validCIDR {
+				return nil, err
+			}
+
+			reservedIPRange, err := s.reserveIPV4Range(ctx, newFiler, reservedIPV4CIDR)
+			if err != nil {
+				return nil, err
+			}
+
+			// Adding the reserved IP range to the instance object
+			newFiler.Network.ReservedIpRange = reservedIPRange
+
+			// Possible cases are 1) CreateInstanceAborted, 2)CreateInstance running in background
+			// In either cases, the listInstances responses would consists of the updated result
+			defer s.config.ipAllocator.Release(reservedIPRange)
```

Comment:
The defer should go right after you reserve the IP.  That way it gets invoked if the function exits for any reason afterwards.